cloud-config: honor cloud_dir setting

aciba90 opened the pull request:
## Proposed Commit Message
<!-- Include a proposed commit message because all PRs are squash merged -->

```
cloud-config: honor cloud_dir setting

- Modules affected: cmd.main, apport, devel.logs (collect-logs),
cc_snap, sources.DataSourceAzure, sources.DataSourceBigstep,
util:fetch_ssl_details.
- testing: Extend and port to pytest unit tests, add integration test.

LP: #1976564
```

## Additional Context
<!-- If relevant -->
https://bugs.launchpad.net/cloud-init/+bug/1976564
https://warthogs.atlassian.net/browse/SC-1085

## Test Steps
<!-- Please include any steps necessary to verify (and reproduce if
this is a bug fix) this change on a live deployed system,
including any necessary configuration files, user-data,
setup, and teardown. Scripts used may be attached directly to this PR. -->
```sh
cat > /tmp/95-custom-cloud-dir.cfg <<EOF
system_info:
  paths:
    cloud_dir: /new-cloud-dir
EOF

CONTAINER_NAME=dev-k
lxc launch ubuntu-daily:kinetic $CONTAINER_NAME
lxc file push /tmp/95-custom-cloud-dir.cfg $CONTAINER_NAME/etc/cloud/cloud.cfg.d/
lxc exec $CONTAINER_NAME -- rm -rf /var/lib/cloud
lxc exec $CONTAINER_NAME -- cloud-init clean --logs --reboot
lxc exec $CONTAINER_NAME -- cloud-init status --wait --long

# Expect no logs referencing use of /var/lib/cloud
lxc exec $CONTAINER_NAME -- grep /var/lib /var/log/cloud-init.log

# Expect logs referencing use of new cloud-dir
lxc exec $CONTAINER_NAME -- grep /new-cloud-dir var/log/cloud-init.log

# Expect to find no files in /var/lib/cloud
lxc exec $CONTAINER_NAME find /var/lib/cloud

# Expect user-data file pointing to the correct cloud-dir
lxc exec $CONTAINER_NAME -- cloud-init collect-logs -h
```

## Checklist:
<!-- Go over all the following points, and put an `x` in all the boxes
that apply. -->
 - [x] My code follows the process laid out in [the documentation](https://cloudinit.readthedocs.io/en/latest/topics/contributing.html)
 - [x] I have updated or added any unit tests accordingly
 - [ ] I have updated or added any documentation accordingly

TheRealFalcon commented:
@aciba90 , I noticed that this PR is in draft state. Is it still in progress, or is it ready for review?

aciba90 commented:
> @aciba90 , I noticed that this PR is in draft state. Is it still in progress, or is it ready for review?

Yes, it is, @TheRealFalcon. I want to run the new integration test on Azure. I will let you know when ready, thanks.

aciba90 commented:
I have tested the added integration test on Azure, and added the Azure mark to execute it there too. This PR is ready to be reviewed @TheRealFalcon, thanks.

aciba90 commented on tests/integration_tests/bugs/test_lp1976564.py:
```diff
@@ -0,0 +1,63 @@
+
+@pytest.mark.lxd_container
+@pytest.mark.lxd_vm
+@pytest.mark.azure
```
The only reason was to only cover the affected components in the bug. But it is better to unrestrict it to cover as much as possible, thanks for the suggestion.

aciba90 commented on tests/integration_tests/bugs/test_lp1976564.py:
```diff
@@ -0,0 +1,63 @@
+    cfg = tmpdir.join(CUSTOM_CLOUD_DIR_FN)
+    with open(cfg, "w") as f:
+        f.write(CUSTOM_CLOUD_DIR)
+    client.push_file(cfg, f"/etc/cloud/cloud.cfg.d/{CUSTOM_CLOUD_DIR_FN}")
```
Thanks, much clearer with your suggestion!

aciba90 commented on cloudinit/sources/DataSourceBigstep.py:
```diff
@@ -31,22 +33,26 @@ def _get_data(self, apply_filter=False):
-    return content
+        return f"metadata ({self._get_url_from_file()})"
+
+    @lru_cache(maxsize=1, typed=False)
```
Done

aciba90 commented on tests/unittests/sources/test_bigstep.py:
```diff
@@ -0,0 +1,46 @@
+import json
```
It was a pleasure.

aciba90 commented:
> Great work here! I left a few comments inline, but nothing major.

Thanks a lot, @TheRealFalcon! I have addressed your comments / suggestions and this PR is ready to be reviewed after the build passes.

TheRealFalcon reviewed (approved):
Thanks!
